feat(site): support deleting dev containers

DanielleMaywood opened the pull request:
Closes https://github.com/coder/coder/issues/19062

Add logic to the frontend to allow deleting Dev Containers

---

PR Stack:
- [x] https://github.com/coder/coder/pull/21247
- [x] https://github.com/coder/coder/pull/21248 
- [ ] https://github.com/coder/coder/pull/21249 (Current PR 🚨)

DanielleMaywood committed:
feat(site): support deleting dev containers

johnstcn reviewed (commented):
Changes look OK to me, but I would defer to someone with more FE experience.

One design question/nit: should we also move the "recreate" button inside the new "more actions"?

DanielleMaywood commented:
> Changes look OK to me, but I would defer to someone with more FE experience.

Added Kayla as a reviewer

> One design question/nit: should we also move the "recreate" button inside the new "more actions"?

Definitely a possibility, although I think "Recreate" is way more likely to be needed than "Delete"? Unsure though on if that changes anything from a design point though

johnstcn commented:
> Definitely a possibility, although I think "Recreate" is way more likely to be needed than "Delete"?

Agreed, it may make sense to revisit this when we have more devcontainer-related actions.

EhabY commented on site/src/modules/resources/AgentDevcontainerCard.tsx:
```diff
@@ -167,7 +161,21 @@ export const AgentDevcontainerCard: FC<AgentDevcontainerCardProps> = ({
+	};
+	const rebuildButtonLabel =
+		statusLabels[devcontainer.status] ??
+		(devcontainer.container === undefined ? "Start" : "Rebuild");
```
Nit: At that point maybe refactor this into a function with a switch (or similar) instead of creating a record to hold two constants.

Same for the other conditions.

EhabY commented on site/src/modules/resources/AgentDevcontainerMoreActions.tsx:
```diff
@@ -0,0 +1,140 @@
+					["agents", parentAgent.id, "containers"],
+					context.previousData,
+				);
+			}
```
Is this the standard way of error handling here, that we just show the previous data without any indication?

DanielleMaywood replied:
Oops this slipped my mind, will add some proper error handling

EhabY reviewed (approved):
Not super familiar with storybook but the rest looks good to me!

aslilac commented on site/src/modules/resources/AgentDevcontainerMoreActions.stories.tsx:
```diff
@@ -0,0 +1,80 @@
+
+		const deleteSpy = spyOn(API, "deleteDevContainer").mockResolvedValue(
+			undefined as never,
+		);
```
what? `undefined as never` is definitely not a cast that makes any sense

DanielleMaywood replied:
Oops, my bad

DanielleMaywood committed:
test: remove useless assertions and type-cast

The assertions are covering the same functionality as our snapshots and
the `unknown as never` just doesn't make any sense.

DanielleMaywood committed:
refactor: extract rebuild button label logic into a helper

DanielleMaywood committed:
refactor: introduce queryKey helper

replace error-prone hardcoded query key with the helper

DanielleMaywood committed:
refactor: use `workspaceAgentContainersKey` more, and drop `useEffect`

Remove a (potentially) unneeded `useEffect` and make use of
`workspaceAgentContainersKey` in another place.

DanielleMaywood committed:
refactor: use `workspaceAgentContainersKey` in more places

DanielleMaywood committed:
chore: appease the formatter gods

aslilac commented on site/src/api/queries/workspaces.ts:
```diff
@@ -490,3 +490,9 @@ export const workspaceAgentCredentials = (
 	};
 };
+
+export const workspaceAgentContainersKey = (agentId: string) => [
```
there's still no actual query that uses this key function. and this is the only usage of the string `"containers"` in this file, so something is definitely smelly.

we usually define these functions directly above the query function that uses it. it needs to be used by both the query function that specifies the key _and_ any places that try to cancel/mutate/refetch/whatever that query. if we don't use it everywhere it's not actually doing it's job of making sure things stay in sync.

aslilac commented on site/src/modules/resources/AgentDevcontainerMoreActions.tsx:
```diff
@@ -0,0 +1,136 @@
+
+			return { previousData };
+		},
+		onError: (_, __, context) => {
```
nit: rather than `__` I usually would do something like `_a` and `_b` to make it clearer what the ignored arguments are. not really a hard rule tho, just personal taste.

aslilac commented on site/src/modules/resources/useAgentContainers.ts:
```diff
@@ -13,13 +14,14 @@ export function useAgentContainers(
 		isLoading: queryIsLoading,
 	} = useQuery({
-		queryKey: ["agents", agent.id, "containers"],
+		queryKey,
```
oh it's because the query is in here.

could we move this query to an actual queries file or do we think that would make it prone to misuse? if it needs to stay here, could we colocate the key in this file rather than defining it at a distance?

DanielleMaywood replied:
Yeah I've been pondering on this all day. This `useAgentContainers` flow is currently a bit of a mess anyways and I think it needs a refactor. I'll move the query up a level to where you were expecting it

DanielleMaywood committed:
feat: add error dialog

DanielleMaywood committed:
fix: broken test
